# Modifications to SAM 2
#   upstream: https://github.com/facebookresearch/sam2
#   upstream licence: Apache-2.0
# Modified by the HuRo authors, 2026. This diff is derived from the upstream above and
# is governed by that licence -- NOT by the Apache-2.0 in ../LICENSE, which covers only
# HuRo's own code. See ../THIRD_PARTY_NOTICES.md.

diff --git a/sam2/sam2_video_predictor.py b/sam2/sam2_video_predictor.py
index 5a7e1a0..5184aa2 100644
--- a/sam2/sam2_video_predictor.py
+++ b/sam2/sam2_video_predictor.py
@@ -580,7 +580,8 @@ class SAM2VideoPredictor(SAM2Base):
             )
             processing_order = range(start_frame_idx, end_frame_idx + 1)
 
-        for frame_idx in tqdm(processing_order, desc="propagate in video"):
+        #for frame_idx in tqdm(processing_order, desc="propagate in video"):
+        for frame_idx in processing_order:
             pred_masks_per_obj = [None] * batch_size
             for obj_idx in range(batch_size):
                 obj_output_dict = inference_state["output_dict_per_obj"][obj_idx]
diff --git a/sam2/utils/misc.py b/sam2/utils/misc.py
index b65ee82..47112e7 100644
--- a/sam2/utils/misc.py
+++ b/sam2/utils/misc.py
@@ -204,11 +204,73 @@ def load_video_frames(
             async_loading_frames=async_loading_frames,
             compute_device=compute_device,
         )
+    elif isinstance(video_path, list): # custom
+        return load_video_frames_from_numpy_list(
+            video_frames=video_path, # list[np.ndarray], (H,W,3), BGR
+            image_size=image_size,
+            offload_video_to_cpu=offload_video_to_cpu,
+            img_mean=img_mean,
+            img_std=img_std,
+            compute_device=compute_device,
+        )
     else:
         raise NotImplementedError(
             "Only MP4 video and JPEG folder are supported at this moment"
         )
 
+def load_video_frames_from_numpy_list(
+    video_frames, # list of np.ndarray, each (H,W,3), uint8 BGR
+    image_size,
+    offload_video_to_cpu,
+    img_mean=(0.485, 0.456, 0.406),
+    img_std=(0.229, 0.224, 0.225),
+    compute_device=torch.device("cuda"),
+):
+    if not isinstance(video_frames, list) or len(video_frames) == 0:
+        raise RuntimeError("video_frames must be a non-empty list of np.ndarray")
+    first = video_frames[0]
+    if not isinstance(first, np.ndarray):
+        raise TypeError(f"Expected list of np.ndarray, got {type(first)}")
+    if first.ndim != 3 or first.shape[2] != 3:
+        raise ValueError(f"Expected frame shape (H, W, 3), got {first.shape}")
+    if first.dtype != np.uint8:
+        raise TypeError(f"Expected uint8 frames, got {first.dtype}")
+
+    video_height, video_width = first.shape[0], first.shape[1]
+    N = len(video_frames)
+    #print('num_frames:', N)
+    frames_bgr = np.stack(video_frames, axis=0) # (N, H, W, 3)
+    frames_rgb = frames_bgr[..., ::-1] # BGR→RGB
+    frames_f32 = frames_rgb.astype(np.float32) / 255.0 # normalize 0~1
+    images = torch.from_numpy(frames_f32).permute(0,3,1,2)  # torch (N,3,H,W), CPU
+
+    assert not offload_video_to_cpu
+    if video_height == video_width == image_size:
+        images = images.to(compute_device, non_blocking=True)
+    else:
+        div = 2
+        chunk_size = (N + div - 1) // div
+        out = torch.empty((N, 3, image_size, image_size), device=compute_device, dtype=images.dtype)
+        for start in range(0, N, chunk_size):
+            end = min(start + chunk_size, N)
+            chunk = images[start:end].to(compute_device, non_blocking=True)
+            chunk = torch.nn.functional.interpolate(
+                chunk,
+                size=(image_size, image_size),
+                mode="nearest",
+            )
+            out[start:end] = chunk
+        images = out
+
+    img_mean_t = torch.tensor(img_mean, dtype=torch.float32)[:, None, None]  # (3,1,1)
+    img_std_t  = torch.tensor(img_std , dtype=torch.float32)[:, None, None]
+    if not offload_video_to_cpu:
+        img_mean_t = img_mean_t.to(compute_device)
+        img_std_t  = img_std_t.to(compute_device)
+
+    images -= img_mean_t
+    images /= img_std_t
+    return images, video_height, video_width
 
 def load_video_frames_from_jpg_images(
     video_path,
diff --git a/setup.py b/setup.py
index 78a634c..5cb9f6a 100644
--- a/setup.py
+++ b/setup.py
@@ -22,8 +22,8 @@ with open("README.md", "r", encoding="utf-8") as f:
 
 # Required dependencies
 REQUIRED_PACKAGES = [
-    "torch>=2.5.1",
-    "torchvision>=0.20.1",
+    "torch>=2.1.0",
+    "torchvision>=0.16.0",
     "numpy>=1.24.4",
     "tqdm>=4.66.1",
     "hydra-core>=1.3.2",
